tsdb/chunkenc/histo.go

```diff
@@ -0,0 +1,711 @@
+		copy(buf, c.b.stream)
+		c.b.stream = buf
+	}
+}
+
+// Appender implements the Chunk interface.
+func (c *HistoChunk) Appender() (Appender, error) {
+	it := c.iterator(nil)
+
+	// To get an appender we must know the state it would have if we had
+	// appended all existing data from scratch.
+	// We iterate through the end and populate via the iterator's state.
+	for it.Next() {
+	}
+	if err := it.Err(); err != nil {
+		return nil, err
+	}
+
+	a := &histoAppender{
+		b: &c.b,
+
+		schema:   c.schema,
+		posSpans: c.posSpans,
+		negSpans: c.negSpans,
+
+		t:               it.t,
+		cnt:             it.cnt,
+		zcnt:            it.zcnt,
+		tDelta:          it.tDelta,
+		cntDelta:        it.cntDelta,
+		zcntDelta:       it.zcntDelta,
+		posbuckets:      it.posbuckets,
+		negbuckets:      it.negbuckets,
+		posbucketsDelta: it.posbucketsDelta,
+		negbucketsDelta: it.negbucketsDelta,
+
+		sum:      it.sum,
+		leading:  it.leading,
+		trailing: it.trailing,
+
+		buf64: make([]byte, binary.MaxVarintLen64),
+	}
+	if binary.BigEndian.Uint16(a.b.bytes()) == 0 {
+		a.leading = 0xff
+	}
+	return a, nil
+}
+
+// TODO fix this
+func (c *HistoChunk) iterator(it Iterator) *histoIterator {
+	// Should iterators guarantee to act on a copy of the data so it doesn't lock append?
+	// When using striped locks to guard access to chunks, probably yes.
+	// Could only copy data if the chunk is not completed yet.
+	//if histoIter, ok := it.(*histoIterator); ok {
+	//	histoIter.Reset(c.b.bytes())
+	//	return histoIter
+	//}
+	return &histoIterator{
+		// The first 2 bytes contain chunk headers.
+		// We skip that for actual samples.
+		br:       newBReader(c.b.bytes()[2:]),
+		numTotal: binary.BigEndian.Uint16(c.b.bytes()),
+		t:        math.MinInt64,
+	}
+}
+
+// Iterator implements the Chunk interface.
+//func (c *HistoChunk) Iterator(it Iterator) Iterator {
+//	return c.iterator(it)
+//}
+
+type histoAppender struct {
+	b *bstream
+
+	// Meta
+	schema             int32
+	posSpans, negSpans []histogram.Span
+
+	// for the fields that are tracked as dod's
+	t                           int64
+	cnt, zcnt                   uint64
+	tDelta, cntDelta, zcntDelta uint64
+
+	posbuckets, negbuckets           []int64
+	posbucketsDelta, negbucketsDelta []int64 // these could probably be uints instead. if the counts go up, the delta will always be >=0
+
+	// for the fields that are gorilla xor coded
+	sum      float64
+	leading  uint8
+	trailing uint8
+
+	buf64 []byte // for working on varint64's
+}
+
+func putVarint(b *bstream, buf []byte, x int64) {
+	for _, byt := range buf[:binary.PutVarint(buf, x)] {
+		b.writeByte(byt)
+	}
+}
+
+func putUvarint(b *bstream, buf []byte, x uint64) {
+	for _, byt := range buf[:binary.PutUvarint(buf, x)] {
+		b.writeByte(byt)
+	}
+}
+
+// we use this for millisec timestamps and all counts
+// for now this is copied from xor.go - we will probably want to be more conservative (use fewer bits for small values) - can be tweaked later
```

Comment:
The "ideal" bit pattern has already been "found" in the design doc: 3/6/9/12/64.
This is actually important for the outcome of the hackathon, as we want to see how efficient histograms are stored. With the weird bit-buckets from Prometheus, we will not be very efficient.

Reply:
To be precise: For timestamps, the ideal layout might be different. The one mentioned above is derived from production data just for buckets.
But in any case, that bucketing will still be better than the one used right now. Most timestamp-dod's will be either 0 or very small.